Libraries/Opc.Ua.Client/Subscription.cs

```diff
@@ -1958,15 +2006,26 @@ private async Task OnMessageReceived()
                             m_messageCache.AddLast(ii.Value.Message);
                             ii.Value.Processed = true;
 
-                            //Keep the last sequence number processed going up
+                            // Keep the last sequence number processed going up
                             if (ii.Value.SequenceNumber > m_lastSequenceNumberProcessed)
                             {
                                 m_lastSequenceNumberProcessed = ii.Value.SequenceNumber;
+                                m_resyncLastSequenceNumberProcessed = false;
+                            }
+                        }
+
+                        // process keep alive messages
+                        else if (ii.Next == null && ii.Value.Message == null && !ii.Value.Processed)
+                        {
+                            if (keepAliveToProcess == null)
+                            {
+                                keepAliveToProcess = new List<IncomingMessage>();
                             }
+                            keepAliveToProcess.Add(ii.Value);
                         }
 
                         // check for missing messages.
-                        if (ii.Next != null && ii.Value.Message == null && !ii.Value.Processed && !ii.Value.Republished)
+                        else if (ii.Next != null && ii.Value.Message == null && !ii.Value.Processed && !ii.Value.Republished)
```

Comment:
Why was this `else` added? I remember this has to happen after processing the message regardless

Reply:
I figured it can not hit the case when any other case wa already handled and may save some extra checks.